cni/network/network.go

```diff
@@ -457,199 +454,243 @@ func (plugin *netPlugin) Add(args *cniSkel.CmdArgs) error {
+		return nwInfo, err
+	}
 
-		err = plugin.nm.CreateNetwork(&nwInfo)
+	nwDNSInfo, err := getNetworkDNSSettings(nwCfg, result)
+	if err != nil {
+		err = plugin.Errorf("Failed to getDNSSettings: %v", err)
+		return nwInfo, err
+	}
+
+	log.Printf("[cni-net] nwDNSInfo: %v", nwDNSInfo)
+	// Update subnet prefix for multi-tenant scenario
+	if err = updateSubnetPrefix(cnsNetworkConfig, &subnetPrefix); err != nil {
+		err = plugin.Errorf("Failed to updateSubnetPrefix: %v", err)
+		return nwInfo, err
+	}
+
+	// Create the network.
+	nwInfo = network.NetworkInfo{
+		Id:           networkID,
+		Mode:         nwCfg.Mode,
+		MasterIfName: masterIfName,
+		AdapterName:  nwCfg.AdapterName,
+		Subnets: []network.SubnetInfo{
+			{
+				Family:  platform.AfINET,
+				Prefix:  subnetPrefix,
+				Gateway: gateway,
+			},
+		},
+		BridgeName:                    nwCfg.Bridge,
+		EnableSnatOnHost:              nwCfg.EnableSnatOnHost,
+		DNS:                           nwDNSInfo,
+		Policies:                      policies,
+		NetNs:                         args.Netns,
+		DisableHairpinOnHostInterface: nwCfg.DisableHairpinOnHostInterface,
+		IPV6Mode:                      nwCfg.IPV6Mode,
+		ServiceCidrs:                  nwCfg.ServiceCidrs,
+	}
+
+	nwInfo.IPAMType = nwCfg.Ipam.Type
+
+	if len(result.IPs) > 0 {
+		var podnetwork *net.IPNet
+		_, podnetwork, err = net.ParseCIDR(result.IPs[0].Address.String())
 		if err != nil {
-			err = plugin.Errorf("Failed to create network: %v", err)
-			return err
+			return nwInfo, fmt.Errorf("%w", err)
 		}
 
-		log.Printf("[cni-net] Created network %v with subnet %v.", networkId, subnetPrefix.String())
-	} else {
-		if !nwCfg.MultiTenancy {
-			// Network already exists.
-			log.Printf("[cni-net] Found network %v with subnet %v.", networkId, nwInfo.Subnets[0].Prefix.String())
-			result, resultV6, err = plugin.ipamInvoker.Add(nwCfg, args, &subnetPrefix, nwInfo.Options)
-			if err != nil {
-				return err
-			}
+		nwInfo.PodSubnet = network.SubnetInfo{
+			Family:  platform.GetAddressFamily(&result.IPs[0].Address.IP),
+			Prefix:  *podnetwork,
+			Gateway: result.IPs[0].Gateway,
+		}
+	}
 
-			nwInfo.IPAMType = nwCfg.Ipam.Type
+	nwInfo.Options = options
+	setNetworkOptions(cnsNetworkConfig, &nwInfo)
 
-			defer func() {
-				if err != nil {
-					if result != nil && len(result.IPs) > 0 {
-						if er := plugin.ipamInvoker.Delete(&result.IPs[0].Address, nwCfg, args, nwInfo.Options); er != nil {
-							err = plugin.Errorf("Failed to cleanup when NwInfo was nil with error %v, after Add failed with error %w", er, err)
-						}
-					}
-					if resultV6 != nil && len(resultV6.IPs) > 0 {
-						if er := plugin.ipamInvoker.Delete(&resultV6.IPs[0].Address, nwCfg, args, nwInfo.Options); er != nil {
-							err = plugin.Errorf("Failed to cleanup when NwInfo was nil with error %v, after Add failed with error %w", er, err)
-						}
-					}
-				}
-			}()
-		}
+	addNatIPV6SubnetInfo(nwCfg, resultV6, &nwInfo)
```

Comment:
perhaps subject to a larger refactor, but the v6 codepath is very coupled to v4, this is going to be called on every v4 invocation